TCMManagement/BusinessLayer/AppointmentService.cs

```diff
@@ -1,85 +1,149 @@
-
-                if(isPatient)
-                    return context.Appointments.Where(a => a.PatientId == id ).ToList();
-                else
-                    return context.Appointments.Where(a => a.PersonId == id ).ToList(); 
-            }
-            return context.Appointments.ToList();
-        }
-
-        public Appointment GetItemById(int id)
-        {
-            return context.Appointments.FirstOrDefault(a => a.AppointmentId == id);
-        }
-
-        public Appointment SearchItem(string s)
-        {
-            return null;
-        }
-
-        public bool UpdateItem(int id, Appointment a)
-        {
-            return true;
-        }
-
-        public bool DeleteItem(int id)
-        {
-            Appointment a = context.Appointments.Find(id);
-            if (a == null)
-            {
-                return false;
-            }
-            context.Appointments.Remove(a);
-            SaveChanges();
-            return true;
-        }
-
-        public void MarkAsModified(Appointment item)
-        {
-            context.Entry(item).State = EntityState.Modified;
-        }
-
-        public int SaveChanges()
-        {
-            return context.SaveChanges();
-        }
-
-        public void Dispose()
-        {
-            context.Dispose();
-        }
-    }
+using System.Collections.Generic;
+using System.Linq;
+using TCMManagement.Models;
+using System.Data.Entity;
+using static TCMManagement.BusinessLayer.Constants;
+using TCMManagement.BusinessLayer;
+using System;
+using System.Web.Http.OData;
+
+namespace TCMManagement.BusinessLayer
+{
+    public class AppointmentService : IEntityServices<Appointment>
+    {
+        private TcmContext context;
+
+        public AppointmentService()
+        {
+            context = new TcmContext();
+        }
+
+        public AppointmentService(TcmContext inputContext)
+        {
+            context = inputContext;
+        }
+
+        // NOTE: We might need to check the validity of this appointment.
+        public Appointment CreateItem(Appointment a)
+        {
+            // check the start time of the new appointment
+            var conflictList = context.Appointments
+                                .Where(ea => ea.PatientId == a.PatientId || ea.PersonId == a.PersonId)
+                                .ToList();
+            foreach (var conflictAppointment in conflictList)
+            {
+                // any time overlap will cause conflict
+                if ((conflictAppointment.TimeStart >= a.TimeStart && conflictAppointment.TimeStart < a.TimeEnd)
+                 || (conflictAppointment.TimeEnd > a.TimeStart && conflictAppointment.TimeEnd <= a.TimeEnd))
+                {
+                    return null;
+                }
+            }
+            // no conflict, add the appointment
+            context.Appointments.Add(a);
+            SaveChanges();
+            return context.Appointments.ToList().Last();

```

Comment:
Any idea what sorting rule is used to convert the table column to list ? In unit test, looks like the last appointment is not always the new added one.